File: 프로그래머스/lv3/43238. 입국심사/입국심사.py
def solution(n, times):
    answer = 0
    # mid는 mid시간 동안 n명의 사람들이
    #  심사를 모두 마칠 수 있는 시간인지 판단 하는 기준 
    l = min(times)
    r = max(times)*n

    while l <= r:
        mid = (l+r) // 2

        count = 0
        for i in times:
            count += mid // i
            if count >= n:
                break

        if count >= n:
            answer = mid
            r = mid - 1
        elif count < n:
            l = mid + 1
            
    return answer

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    public long solution(int n, int[] times) {
        long answer = 0;
        Arrays.sort(times);
        long l = 0;
        long r = (long)times[times.length-1] * n;
        
        while(l <= r){
            // 해당 시간 동안 각 심사대별로 몇 명을 처리할 수 있는지 
            long mid = (l + r) / 2;
            long count = 0;
            for(int t : times){
                count += mid / t;
            }
            
            // 더 많은 사람들을 처리할 수 있음.
            if(count >= n){
                r = mid - 1;
                answer = mid;
            }else{
                l = mid + 1;
            }
            
        }
        
        return answer;
    }
}